### ISSUE TITLE
Add a scrapy wrapper for `twisted.internet.reactor` imports

### ISSUE THREAD

**Issue by nyov** [scrapy/scrapy#4435]
I think we should add a scrapy replacement for `twisted.internet.reactor` imports.
(Also see https://github.com/scrapy/scrapy/issues/4401#issuecomment-596111817 .)

#4406 fixed the current issues with wrong order of reactor imports, but it looks ugly to import `reactor` in all those methods.
I would rather we could revert this change and replace it with a scrapy utility function to lazily(?) import reactor there. Some kind of singleton provider, which I think should also replace the `install_reactor()` function, to have a single reactor provider (which can be used by 3rd party code).

I'm not quite sure if it needs to be lazy, I don't think we need to defer grabbing the reactor until `scrapy/crawler.py` wants it, just long enough to have all the settings loaded, to know which one we want?
Easiest for 3rd party code might be to mock-patch `twisted.internet.reactor`, but I don't know if that's the right approach.

**Comment by Gallaecio** [comment]
We could have something like:

```python
def get_reactor():
    from twisted.internet import reactor
    return reactor
```

So that we can import that function, and use `get_reactor()` where we currently use `reactor`.

**Comment by faizan2700** [comment]
I am working on this. @nyov  can you please tell me where is install_reactor() function I did not see it in mentioned pull request. 
I think scrapy function which provides the singleton object of reactor needs to implemented ( what else is required? ).

**Comment by wRAR** [comment]
Ideally this should be implemented in Twisted, not in Scrapy...

**Comment by nyov** [comment]
> Ideally this should be implemented in Twisted, not in Scrapy...

That would certainly make it easier. But I'm not fully convinced it's twisted's problem, since the [twisted reactor module][1] says:

> New application code should prefer to pass and accept the reactor as a
> parameter where it is needed, rather than relying on being able to import this
> module to get a reference.

This, to me, encourages not relying on `twisted.internet.reactor` to actually _be_ the reactor or a reference to it, even if it's not actually a deprecated interface per se. 

To me, it's a slightly weird interface, given that the `twisted.internet.reactor` module gets replaced with the reactor instance. (By this code in `twisted.internet.main.installReactor`)
```py
if 'twisted.internet.reactor' in sys.modules:
    raise error.ReactorAlreadyInstalledError("reactor already installed")
twisted.internet.reactor = reactor
sys.modules['twisted.internet.reactor'] = reactor
```

In [`twisted.internet.reactor`][1], all it does (before it becomes the reactor instance) is
```py
import sys
del sys.modules['twisted.internet.reactor']
from twisted.internet import default
default.install()
```
and [`default.install()`][2] only chooses between `select`, `poll` and `epoll`, depending on estimated platform support. (So we already need a non-default [`install_reactor()`][3] to switch to asyncioreactor, in any event.)

**But** we don't need to evaluate this `default` stuff at all, when our `TWISTED_REACTOR` setting is set. So what currently happens, importing the default reactor via `twisted.internet.reactor`, only to suppress the `ReactorAlreadyInstalledError` and re-install a custom reactor afterwards, is arguably the wrong order?

Instead we might write our own replacement for the whole "installer" codebase that's strewn across `t.i.reactor`, `t.i.default`, and `t.i.main`. Since we circumvent most of that code with the setting `TWISTED_REACTOR` and [`install_reactor()`][3] anyway, we've essentially already done that, in a crude way (blowing away the current reactor, if any). And we could fall back to the `default.install()` routine only after a custom reactor isn't set via `TWISTED_REACTOR` (or emulate the same in our own `default_settings`, by pre-setting `TWISTED_REACTOR` depending on platform)?

[1]: https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/reactor.py
[2]: https://github.com/twisted/twisted/blob/trunk/src/twisted/internet/default.py
[3]: https://github.com/scrapy/scrapy/blob/6c747953f97d60e5668cff1af3f091979ceaaa83/scrapy/utils/reactor.py#L52

**Comment by AdrianAtZyte** [comment]
Now that we support not running a Twisted reactor at all, I think we have methods to replace all the things you might need the reactor for before. Maybe we can close this now?
